packages/x-date-pickers/src/internals/hooks/usePickerState.ts

```diff
@@ -41,127 +79,158 @@ interface PickerStateProps<TInput, TDateValue> {
 
-  const utils = useUtils<TDateValue>();
+  const utils = useUtils<TDate>();
   const { isOpen, setIsOpen } = useOpenState(props);
 
-  function initDraftableDate(date: TDateValue): Draftable<TDateValue> {
-    return { committed: date, draft: date };
-  }
-
   const parsedDateValue = React.useMemo(
     () => valueManager.parseInput(utils, value),
     [valueManager, utils, value],
   );
 
-  const [lastValidDateValue, setLastValidDateValue] = React.useState<TDateValue | null>(
-    parsedDateValue,
-  );
-
-  React.useEffect(() => {
-    if (parsedDateValue != null) {
-      setLastValidDateValue(parsedDateValue);
-    }
-  }, [parsedDateValue]);
-
-  const [draftState, dispatch] = React.useReducer(
-    (state: Draftable<TDateValue>, action: DraftAction<TDateValue>): Draftable<TDateValue> => {
-      switch (action.type) {
-        case 'reset':
-          return initDraftableDate(action.payload);
-        case 'update':
-          return {
-            ...state,
-            draft: action.payload,
-          };
-        default:
-          return state;
-      }
-    },
-    parsedDateValue,
-    initDraftableDate,
-  );
-  if (!valueManager.areValuesEqual(utils, draftState.committed, parsedDateValue)) {
-    dispatch({ type: 'reset', payload: parsedDateValue });
-  }
+  const [lastValidDateValue, setLastValidDateValue] = React.useState<TDate | null>(parsedDateValue);
 
-  const [initialDate, setInitialDate] = React.useState<TDateValue>(draftState.committed);
+  const [dateState, setDateState] = React.useState<DateState<TDate>>(() => ({
+    committed: parsedDateValue,
+    draft: parsedDateValue,
+    resetTarget: parsedDateValue,
+  }));
 
-  // Mobile keyboard view is a special case.
-  // When it's open picker should work like closed, because we are just showing text field
-  const [isMobileKeyboardViewOpen, setMobileKeyboardViewOpen] = React.useState(false);
+  const setDate = React.useCallback(
+    (params: DateStateAction<TDate>) => {
+      setDateState((prev) => {
+        switch (params.action) {
+          case 'setAll':
+          case 'acceptAndClose': {
+            return { draft: params.value, committed: params.value, resetTarget: params.value };
+          }
+          case 'setCommitted': {
+            return { ...prev, draft: params.value, committed: params.value };
+          }
+          case 'setDraft': {
+            return { ...prev, draft: params.value };
+          }
+          default: {
+            return prev;
+          }
+        }
+      });
 
-  const acceptDate = React.useCallback(
-    (acceptedDate: TDateValue, needClosePicker: boolean) => {
-      onChange(acceptedDate);
+      if (
+        !params.internal &&
+        !valueManager.areValuesEqual(utils, dateState.committed, params.value)
+      ) {
+        onChange(params.value);
+      }
 
-      if (needClosePicker) {
+      if (params.action === 'acceptAndClose') {
         setIsOpen(false);
```

Comment:
Are you aware that it doesn't close after selecting the dates in https://deploy-preview-4408--material-ui-x.netlify.app/x/react-date-pickers/date-range-picker/?

`ClickAwayListener` tries to restore the focus after closing, but the `onFocus` on the input makes it to open again.

Reply:
Yes

It is https://github.com/mui/mui-x/issues/4439
I will try to fix it after :+1: